fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleFactory.java

```diff
@@ -18,301 +18,168 @@
+import com.mojang.serialization.Codec;
 
 import net.minecraft.resource.featuretoggle.FeatureSet;
-import net.minecraft.server.MinecraftServer;
-import net.minecraft.world.GameRules;
+import net.minecraft.server.dedicated.management.dispatch.GameRuleType;
+import net.minecraft.world.Category;
+import net.minecraft.world.Visitor;
+import net.minecraft.world.rule.GameRule;
+import net.minecraft.world.rule.GameRules;
 
-import net.fabricmc.fabric.api.gamerule.v1.rule.DoubleRule;
-import net.fabricmc.fabric.api.gamerule.v1.rule.EnumRule;
-import net.fabricmc.fabric.impl.gamerule.EnumRuleType;
-import net.fabricmc.fabric.impl.gamerule.rule.BoundedIntRule;
-import net.fabricmc.fabric.mixin.gamerule.GameRulesBooleanRuleAccessor;
+import net.fabricmc.fabric.impl.gamerule.RuleCategoryExtensions;
+import net.fabricmc.fabric.impl.gamerule.RuleTypeExtensions;
+import net.fabricmc.fabric.impl.gamerule.rpc.FabricGameRuleType;
 
 /**
  * A utility class containing factory methods to create game rule types.
  * A game rule is a persisted, per server data value which may control gameplay aspects.
  *
- * <p>Some factory methods allow specification of a callback that is invoked when the value of a game rule has changed.
- * Typically, the callback is used for game rules which may influence game logic, such as {@link GameRules#DISABLE_RAIDS disabling raids}.
- *
- * <p>To register a game rule, you can use {@link GameRuleRegistry#register(String, GameRules.Category, GameRules.Type)}.
- * For example, to register a game rule that is an integer where the acceptable values are between 0 and 10, one would use the following:
+ * <p>To register a game rule, you can use {@link GameRuleRegistry#register(String, GameRule)} or any register method from {@link GameRules}.
+ * For example, to register a game rule that is an integer where the default value is 1 and the acceptable values are between 0 and 10, one would use the following:
  * <blockquote><pre>
- * public static final GameRules.Key&lt;GameRules.IntRule&gt; EXAMPLE_INT_RULE = GameRuleRegistry.register("exampleIntRule", GameRules.Category.UPDATES, GameRuleFactory.createIntRule(1, 10));
+ * public static final GameRule&lt;Integer&gt; EXAMPLE_INT_RULE = GameRules.registerIntRule("example_int_rule", Category.MISC, 1, 0, 10);
  * </pre></blockquote>
+ * Please note that all register methods in {@link GameRules} internally construct and register the rule.
  *
- * <p>To register a game rule in a custom category, {@link GameRuleRegistry#register(String, CustomGameRuleCategory, GameRules.Type)} should be used.
+ * <p>To register a game rule in a custom category, {@link GameRuleRegistry#register(String, GameRule, CustomGameRuleCategory)} should be used.
+ * Alternatively, cast the {@link GameRule} to {@link RuleCategoryExtensions} and call {@link RuleCategoryExtensions#fabric_setCustomCategory(CustomGameRuleCategory)}.
  *
  * @see GameRuleRegistry
  */
 public final class GameRuleFactory {
```

Comment:
If we are completely revamping the class, I think a redesign might be better. Why not a builder, for example?

Also rename the class since this is not a real factory (factory in Java is a functional interface for creating an instance)